third_party/move/move-model/src/ty.rs

```diff
@@ -2112,3 +2115,123 @@ impl fmt::Display for PrimitiveType {
         }
     }
 }
+
+/// Infers the (conditional) abilities of a struct
+pub fn inst_abilities(
+    struct_abilities: AbilitySet,
+    insts_abilities_meet: AbilitySet,
+) -> AbilitySet {
+    if struct_abilities.has_ability(Ability::Key)
+        && insts_abilities_meet.has_ability(Ability::Store)
+    {
+        struct_abilities
+            .intersect(insts_abilities_meet)
+            .add(Ability::Key)
+    } else {
+        struct_abilities
+            .intersect(insts_abilities_meet)
+            .remove(Ability::Key)
+    }
+}
+
+pub fn check_struct_inst<F, G, H>(
+    mid: ModuleId,
+    sid: StructId,
+    insts: &[Type],
+    loc: Option<&Loc>,
+    ty_param_abilities: F,
+    get_abilities: G,
+    on_err: H,
+    check: bool,
+) -> AbilitySet
+where
+    F: Fn(u16) -> AbilitySet + Copy,
+    G: Fn(ModuleId, StructId) -> (Vec<AbilitySet>, AbilitySet) + Copy,
+    H: Fn(&Loc, &str) -> () + Copy,
+{
+    let (generic_constraints, struct_abilities) = get_abilities(mid, sid);
+    let insts_abilities_meet = insts
+        .iter()
+        .zip(generic_constraints)
+        .map(|(inst_ty, constraints)| {
+            let inst_abilities = check_instantiation(
```

Comment:
Don't use the same name for this local and the function `inst_abilities` defined above and used below.  Please think of better names for them.  Btw, in a compiler, when I hear "inst" I think "instruction" so spelling out "instantiate" is not a bad idea.

Reply:
Agree with Brian in general but please note that in my world of compilers 'inst' is frequently used for 'instantiation', and so it is in the existing code base, in multiple places.

Reply:
done